### PR TITLE
Fix a couple of compile-fail tests

### PR DESCRIPTION
**Pull request by bjorn3**
```
failures:

---- [compile-fail] compile-fail/match_char.rs stdout ----

error: compile-fail test compiled successfully!
status: exit code: 0
command: "target/debug/miri" "tests/compile-fail/match_char.rs" "-L" "/tmp/compiletestpMJMuw" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/tmp/compiletestpMJMuw/match_char.stage-id" "--sysroot" "/home/bjorn/.xargo/HOST" "-Zmir-emit-validate=1" "-L" "/tmp/compiletestpMJMuw/match_char.stage-id.aux" "-A" "unused"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------

------------------------------------------

thread '[compile-fail] compile-fail/match_char.rs' panicked at 'explicit panic', /home/bjorn/.cargo/registry/src/github.com-1ecc6299db9ec823/compiletest_rs-0.3.11/src/runtest.rs:2544:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- [compile-fail] compile-fail/memleak.rs stdout ----

error: compile-fail test compiled successfully!
status: exit code: 0
command: "target/debug/miri" "tests/compile-fail/memleak.rs" "-L" "/tmp/compiletestpMJMuw" "--target=x86_64-unknown-linux-gnu" "-C" "prefer-dynamic" "-o" "/tmp/compiletestpMJMuw/memleak.stage-id" "--sysroot" "/home/bjorn/.xargo/HOST" "-Zmir-emit-validate=1" "-L" "/tmp/compiletestpMJMuw/memleak.stage-id.aux" "-A" "unused"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------

------------------------------------------

thread '[compile-fail] compile-fail/memleak.rs' panicked at 'explicit panic', /home/bjorn/.cargo/registry/src/github.com-1ecc6299db9ec823/compiletest_rs-0.3.11/src/runtest.rs:2544:9

---- [compile-fail] compile-fail/memleak_rc.rs stdout ----

error: compile-fail test compiled successfully!
status: exit code: 0
command: "target/debug/miri" "tests/compile-fail/memleak_rc.rs" "-L" "/tmp/compiletestpMJMuw" "--target=x86_64-unknown-linux-gnu" "-C" "prefer-dynamic" "-o" "/tmp/compiletestpMJMuw/memleak_rc.stage-id" "--sysroot" "/home/bjorn/.xargo/HOST" "-Zmir-emit-validate=1" "-L" "/tmp/compiletestpMJMuw/memleak_rc.stage-id.aux" "-A" "unused"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------

------------------------------------------

thread '[compile-fail] compile-fail/memleak_rc.rs' panicked at 'explicit panic', /home/bjorn/.cargo/registry/src/github.com-1ecc6299db9ec823/compiletest_rs-0.3.11/src/runtest.rs:2544:9

---- [compile-fail] compile-fail/overflowing-rsh-2.rs stdout ----

error: compile-fail test compiled successfully!
status: exit code: 0
command: "target/debug/miri" "tests/compile-fail/overflowing-rsh-2.rs" "-L" "/tmp/compiletestpMJMuw" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/tmp/compiletestpMJMuw/overflowing-rsh-2.stage-id" "--sysroot" "/home/bjorn/.xargo/HOST" "-Zmir-emit-validate=1" "-L" "/tmp/compiletestpMJMuw/overflowing-rsh-2.stage-id.aux" "-A" "unused"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------

------------------------------------------

thread '[compile-fail] compile-fail/overflowing-rsh-2.rs' panicked at 'explicit panic', /home/bjorn/.cargo/registry/src/github.com-1ecc6299db9ec823/compiletest_rs-0.3.11/src/runtest.rs:2544:9

---- [compile-fail] compile-fail/static_memory_modification.rs stdout ----

error: compile-fail test compiled successfully!
status: exit code: 0
command: "target/debug/miri" "tests/compile-fail/static_memory_modification.rs" "-L" "/tmp/compiletestpMJMuw" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/tmp/compiletestpMJMuw/static_memory_modification.stage-id" "--sysroot" "/home/bjorn/.xargo/HOST" "-Zmir-emit-validate=1" "-L" "/tmp/compiletestpMJMuw/static_memory_modification.stage-id.aux" "-A" "unused"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------

------------------------------------------

thread '[compile-fail] compile-fail/static_memory_modification.rs' panicked at 'explicit panic', /home/bjorn/.cargo/registry/src/github.com-1ecc6299db9ec823/compiletest_rs-0.3.11/src/runtest.rs:2544:9


failures:
    [compile-fail] compile-fail/match_char.rs
    [compile-fail] compile-fail/memleak.rs
    [compile-fail] compile-fail/memleak_rc.rs
    [compile-fail] compile-fail/overflowing-rsh-2.rs
    [compile-fail] compile-fail/static_memory_modification.rs

test result: FAILED. 59 passed; 5 failed; 14 ignored; 0 measured; 0 filtered out
```

Edit: updated failure summary

### CODE REVIEW

File: tests/compile-fail/never_transmute_humans.rs [hunk 1]
```diff
@@ -1,17 +0,0 @@
-// This should fail even without validation
```
Comment (bjorn3) [inline tests/compile-fail/never_transmute_humans.rs]: This causes an ICE

File: tests/compile-fail/reallocate-bad-alignment-2.rs [hunk 2]
```diff
@@ -1,16 +0,0 @@
-#![feature(alloc, allocator_api)]
```
Comment (bjorn3) [inline tests/compile-fail/reallocate-bad-alignment-2.rs]: This is simply not possible with the new api

File: tests/compile-fail/reallocate-bad-alignment.rs [hunk 3]
```diff
@@ -1,16 +0,0 @@
-#![feature(alloc, allocator_api)]
```
Comment (bjorn3) [inline tests/compile-fail/reallocate-bad-alignment.rs]: This is simply not possible with the new api

File: tests/compile-fail/repeat2.rs [hunk 4]
```diff
@@ -1,3 +1,5 @@
+// ignore-test
```
Comment (bjorn3) [inline tests/compile-fail/repeat2.rs]: Takes too long

**Comment by oli-obk** [comment]
how were these not failing before? are they disabled by default?

File: tests/compile-fail/never_transmute_humans.rs (hunk 1 above)
Comment (RalfJung) [reply tests/compile-fail/never_transmute_humans.rs]: That's a bug, isn't it? Why remove the test?

Comment (oli-obk) [reply tests/compile-fail/never_transmute_humans.rs]: it causes an ICE in rustc during compilation, not miri: http://play.integer32.com/?gist=76ba0e33563d046120e629846294f4a8&version=nightly&mode=debug

File: tests/compile-fail/repeat2.rs (hunk 4 above)
Comment (RalfJung) [reply tests/compile-fail/repeat2.rs]: Is it possible to make the array smaller and still get the failure?

If not I'd vote for just removing the test instead.

File: tests/compile-fail/never_transmute_humans.rs (hunk 1 above)
Comment (oli-obk) [reply tests/compile-fail/never_transmute_humans.rs]: still... ignoring it for now is better so we can add it once upstream fixes it

Comment (RalfJung) [reply tests/compile-fail/never_transmute_humans.rs]: Ah, I see.

Is this reported upstream? The ignore-test comment should link to the issue.

File: tests/compile-fail/repeat2.rs (hunk 4 above)
Comment (oli-obk) [reply tests/compile-fail/repeat2.rs]: fun fact: we don't have any limits except stack frame depth anymore. This test won't ever trigger the error

File: tests/compile-fail/never_transmute_humans.rs (hunk 1 above)
Comment (bjorn3) [reply tests/compile-fail/never_transmute_humans.rs]: rust-lang/rust#50570

**Comment by oli-obk** [comment]
Can we turn them back on now? Or what are the next steps for doing that?

**Comment by bjorn3** [comment]
> Can we turn them back on now? Or what are the next steps for doing that?

Once all errors have been converted to the new error format. Sadly compile-test only detects the `constant evaluation error [E0080]` message, but not what caused the error like for example "a memory access tried to interpret some bytes as a pointer".

**Comment by oli-obk** [comment]
you should be able to add `//~^ NOTE a memory access tried to interpret...` in the line after the error

**Comment by bjorn3** [comment]
Don't have time to convert them all right now. Some errors have either strange placement or different message.

File: tests/compile-fail/invalid_enum_discriminant.rs [hunk 5]
```diff
@@ -9,9 +9,10 @@ pub enum Foo {
...
     let f = unsafe { std::mem::transmute::<i32, Foo>(42) };
     match f {
-        Foo::A => {}, //~ ERROR invalid enum discriminant value read
+        Foo::A => {},
         Foo::B => {},
         Foo::C => {},
         Foo::D => {},
     }
-}
+} //~ ERROR constant evaluation error [E0080]
```
Comment (bjorn3) [inline tests/compile-fail/invalid_enum_discriminant.rs]: Why here?

File: tests/compile-fail/match_char.rs [hunk 6]
```diff
@@ -1,6 +1,7 @@
 fn main() {
     assert!(std::char::from_u32(-1_i32 as u32).is_none());
-    match unsafe { std::mem::transmute::<i32, char>(-1) } { //~ERROR tried to interpret an invalid 32-bit value as a char: 4294967295
+    match unsafe { std::mem::transmute::<i32, char>(-1) } { //~ ERROR constant evaluation error [E0080]
```
Comment (bjorn3) [inline tests/compile-fail/match_char.rs]: This test doesn't error

File: tests/compile-fail/overflowing-rsh-2.rs [hunk 7]
```diff
@@ -12,5 +12,6 @@
 
 fn main() {
     // Make sure we catch overflows that would be hidden by first casting the RHS to u32
-    let _n = 1i64 >> (u32::max_value() as i64 + 1); //~ Overflow(Shr)
+    let _n = 1i64 >> (u32::max_value() as i64 + 1); //~ ERROR constant evaluation error [E0080]
```
Comment (bjorn3) [inline tests/compile-fail/overflowing-rsh-2.rs]: This test doesn't error

File: tests/compile-fail/static_memory_modification.rs [hunk 8]
```diff
@@ -3,7 +3,8 @@ static X: usize = 5;
 #[allow(mutable_transmutes)]
 fn main() {
     unsafe {
-        *std::mem::transmute::<&usize, &mut usize>(&X) = 6; //~ ERROR: tried to modify constant memory
+        *std::mem::transmute::<&usize, &mut usize>(&X) = 6; //~ ERROR constant evaluation error [E0080]
```
Comment (bjorn3) [inline tests/compile-fail/static_memory_modification.rs]: Statics aren't marked static yet.

File: src/lib.rs [hunk 9]
```diff
@@ -171,6 +171,7 @@ pub fn eval_main<'a, 'tcx: 'a>(
         Ok(()) => {
             let leaks = ecx.memory().leak_report();
             if leaks != 0 {
+                // TODO: Prevent leaks which aren't supposed to be there
                 //tcx.sess.err("the evaluated program leaked memory");
```
Comment (bjorn3) [inline src/lib.rs]: Needs to be uncommented to allow `memleak.rs` and `memleak_rc.rs` to pass.

File: tests/compiletest.rs [hunk 10]
```diff
@@ -203,7 +203,7 @@ fn run_pass_rustc() {
 }
 
 #[test]
-#[ignore] // TODO: update test errors
+#[should_panic] // TODO: update test errors
```
Comment (bjorn3) [inline tests/compiletest.rs]: Remove when all tests pass

File: tests/compile-fail/reallocate-bad-alignment-2.rs (hunk 2 above)
Comment (eddyb) [reply tests/compile-fail/reallocate-bad-alignment-2.rs]: How so? What's changed?

Comment (bjorn3) [reply tests/compile-fail/reallocate-bad-alignment-2.rs]: `realloc` doesn't have a new alignment param anymore.

```rust
unsafe fn realloc(
    &self, 
    ptr: *mut Opaque, 
    layout: Layout, 
    new_size: usize
) -> *mut Opaque;
```

File: tests/compile-fail/invalid_enum_discriminant.rs (hunk 5 above)
Comment (oli-obk) [reply tests/compile-fail/invalid_enum_discriminant.rs]: Not sure why exactly here, but it's delayed because we don't have validation anymore

**Comment by bjorn3** [comment]
Rebased on top of #376

**Comment by oli-obk** [comment]
can you rebase again?

**Comment by bjorn3** [comment]
Rebased

**Comment by oli-obk** [comment]
Thanks! I'm tracking the leftover failing tests in https://github.com/solson/miri/issues/382

**Pull request merged**